### ISSUE TITLE
Revisit kube-reserved calculation for containerd

### ISSUE THREAD

**Issue by shyamjvs** [awslabs/amazon-eks-ami#1141]
### Background
`kube-reserved` is meant to capture resource reservation for kubernetes system daemons (not including pods). For the amazon EKS ami it's mainly just the kubelet and container-runtime today. `kube-reserved` is typically a function of pod density on the nodes.

The current kube-reserved calculation is 2-3 years old (last updated as part of [this experiment](https://github.com/awslabs/amazon-eks-ami/pull/419) we did) when docker (+ shim) was the only container runtime option in the AMI. However, [Amazon EKS ended support for Docker](https://docs.aws.amazon.com/eks/latest/userguide/dockershim-deprecation.html) starting with the Kubernetes version 1.24 launch. Containerd is the only supported runtime now.

Here's the kube-reserved logic we have in our AMIs today:
https://github.com/awslabs/amazon-eks-ami/blob/6b1df1107d6e7ec2f46612ebb509c16101ded22d/files/bootstrap.sh#L214
https://github.com/awslabs/amazon-eks-ami/blob/6b1df1107d6e7ec2f46612ebb509c16101ded22d/files/bootstrap.sh#L221-L225

### Problem
Per the k8s containerd GA [release blog](https://kubernetes.io/blog/2018/05/24/kubernetes-containerd-integration-goes-ga/#cpu-and-memory) the containerd integration consumes less CPU/memory in comparison with the docker integration (~10-12% - see below). We'll need to repeat the experiment given containerd versions being used in our AMIs are newer (1.5+). But overall we should expect notable savings.

```
At the steady state, with 105 pods, the containerd 1.1 integration consumes less CPU and memory overall compared to
Docker 18.03 CE integration with dockershim. The results vary with the number of pods running on the node, 105 is 
chosen because it is the current default for the maximum number of user pods per node.

As shown in the figures below, compared to Docker 18.03 CE integration with dockershim, the containerd 1.1
integration has 30.89% lower kubelet cpu usage, 68.13% lower container runtime cpu usage, 11.30% lower kubelet 
resident set size (RSS) memory usage, 12.78% lower container runtime RSS memory usage.
```

![image](https://user-images.githubusercontent.com/4333971/210138679-fa2f9c08-ae23-403c-814e-545976f7a0ec.png) | ![image](https://user-images.githubusercontent.com/4333971/210138682-904aaad2-9324-462c-9bf3-fa976594fb22.png)
:-------------------------:|:-------------------------:

### Additional Notes

- With the recent launch of the Prefix Delegation (PD) feature in our AMIs, the supported pod densities have gone up for various instance types. The new logic sets max pod density as 110 for <= 30 vCPU instance-types and 250 for > 30 vCPU instance-types ([xref](https://github.com/awslabs/amazon-eks-ami/blob/3e1c6c2c421be1e3b0b46f5bce7ac28c9b6123fb/files/max-pods-calculator.sh#L145-L159)). Discussed more in [this issue](https://github.com/awslabs/amazon-eks-ami/issues/782). While PD should also be taken into account for kube-reserved calculation, that can happen independent from this change IMO since docker -> containerd switch is a common denominator. The drop in baseline usage will help customers reap incremental benefits, even if it means bumping it up later for PD-specific customers
- Another issue is that - the bottlerocket AMI seems to have copied the same kube-reserved calculation as the EKS AMI (xrefs - [cpu](https://github.com/bottlerocket-os/bottlerocket/blob/develop/sources/api/schnauzer/src/helpers.rs#L1446-L1463), [mem](https://github.com/bottlerocket-os/bottlerocket/blob/develop/sources/api/schnauzer/src/helpers.rs#L1055)) despite using containerd. Once we fix the calculation here, we should go have them update it too

**Comment by shyamjvs** [comment]
@suket22 @mmerkes @natherz97 @jayanthvn @akestner - You've all been close to the ground here. Please add any thoughts you may have.

**Comment by bryanasdev000** [comment]
Related to https://github.com/awslabs/amazon-eks-ami/issues/1145

**Comment by stevehipwell** [comment]
I've just roughly re-run the tests on an AL2 EKS v1.29 node and have the following results.

- Kubelet + containerd use about 65Mb with no pods running
- Each pod with a single container adds around 6Mb to the runtime usage
- 110 pods with 1 container uses about 680MB for the runtime
- 110 pods with 5 containers uses about 1200MB for the runtime

**Comment by sidewinder12s** [comment]
@shyamjvs Any update here/was there any re-evaluation done with Amazon Linux 2023?

**Comment by stevehipwell** [comment]
👀

**Comment by 2rs2ts** [comment]
Hey folks just want to chime in here real quick with the rough outline of what we've been doing at my company running Kubernetes the Hard Way (we also run EKS but our process there works a smidge differently, we're not 100% satisfied with the result but it's pretty good and the below description applies to it as well.)

So basically what we do is separate [system-reserved](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#system-reserved) from [kube-reserved](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#kube-reserved) because they grow at different rates but have different baselines. We have variables in our terraform configs that we use to customize user-data as needed per-environment with both the ability to set a value to a specific amount, and to adjust the growth ratios (more details later.) We use ratios but ultimately perform a ceiling operation and write the values as Mi to avoid those nasty bugs where kubectl reports fractions of bytes. We have such variables for:

* system-reserved and kube-reserved cpu, memory, ephemeral storage
* soft and hard eviction thresholds for memory and nodefs
* eviction minimum reclaim values for memory and nodefs

Now, when it comes to EKS, it's harder to customize the variables, so I think we just use overly-generous defaults (but the same code that sets it all up) for those servers and hope for the best. As our EKS usage ramps up, it might prove untenable...

Anyway, system-reserved's primary growth factors are the burden placed on the server by things like security daemons, non-container logging, SSM, stuff like that. In other words it has a generally pretty low scaling factor, but it starts off at a higher baseline. Meanwhile, kube-reserved's primary growth factors are pod density and pod churn–in particular, we've noticed that clusters with lots of CronJobs that are constantly kicking off, even when staggered, cause the peak resource usage of kubelet and containerd to skyrocket. (On 48xlarges, over 20 cpus being used for just kube-reserved stuff alone is not uncommon.) You *ideally* want to set reserved resources high enough that pods never have to worry about the system being a noisy neighbor, so you have to track growth rates rather than just baselines... though baselines are useful floor values to set for sure rather than using ratios the whole way like we do. (It's just that the instance sizes we use and the reservation ratios we use are big enough that those floors mentioned in previous comments are far below our de facto floors.)

So when we reserve system-reserved resources, we generally use lower ratios but for smaller instance types that ratio bumps up or we just straight up give it a hard override, since that baseline is higher but doesn't grow as fast. For kube-reserved, we use a higher ratio. And FWIW that's ratio of resources available (available cpus, available memory, etc.) But when we reserve ephemeral storage, we first measure how much disk was already used on the instance as it starts up and add that value to our reservation ratios, because things like the size of the OS and programs are kind of a "tax" that you have to pay and won't go up nor down with activity.

Example calculations:

* `reserved cpu = reserved cpu override (if exists) || (available cpus * reserved cpu ratio)`
* `soft eviction threshold mem = soft eviction threshold mem override (if exists)  || (available mem * soft eviction mem ratio)`
* `reserved mem = (reserved mem override (if exists) || (available mem * reserved mem ratio)) + (soft eviction threshold for mem - hard eviction threshold for mem)`

   Regarding this one: it's our understanding that the hard eviction threshold is already taken into account when reserving memory– basically if you tell kubelet to reserve, say, 4GiB, and the hard eviction threshold is 1GiB, it actually reserves 5GiB–but it doesn't take the soft eviction threshold into account, so if you don't add the difference between your soft eviction threshold and the hard eviction threshold onto the reservation, then you are liable to get repeated soft evictions during those pressure scenarios.
* `reserved ephemeral storage = reservered ephemeral storage override (if exists) || (amount of root volume used at time of startup + (root volume size * reserved ephemeral storage ratio))`

And yes we created different cgroups for these two and have monitoring to measure how much resource usage is occurring in said groups, send that to a central monitoring service, and that informs our scaling factors and environment-specific overrides. We found that in order to safely guarantee node stability during high-load scenarios, we had to reserve a LOT more resources than baseline usage in our busier environments.

As far as the ratios go, our values are totally not going to be good references for general purposes, but I will say a few things:

1. when looking at how much disk to reserve, we calculated how big the `/var/log` folder could possibly get if all the logs were rotating every minute, how big the OS and the programs were, and then added some buffer in case we overlooked anything.
2. we had to spend a long time tweaking the ratios and sometimes using hard override values for memory and CPU, and had to aim for around the p75 or so of peak resource usage for kube-reserved because otherwise we'd be reserving half the node's resources which is a big waste
3. we also use the reserved CPUs feature, though we are aware of the long-standing issue where system daemons are not isolated from the cpuset for pods (https://github.com/kubernetes/kubernetes/issues/85764, https://github.com/kubernetes/kubernetes/pull/87452) so this always throws a wrench into our reservation math

I'm sharing this because AFAIK there is no mechanism for doing this built into the EKS AMI already. We customize the AMI to install business-required security daemons anyway so we're doing ok setting up our own stuff (it's inconvenient in some ways, but we make it work) but this seemed like something that the community as a whole would benefit from if it was done by the stock EKS AMI and configurable in similar ways to how we have it.

**Comment by stevehipwell** [comment]
@2rs2ts how do you solve the container scaling problem (see https://github.com/awslabs/amazon-eks-ami/issues/1141#issuecomment-1947048088)?

**Comment by 2rs2ts** [comment]
@stevehipwell as I stated in my post, we use a higher scaling ratio for the kube-reserved (which is the runtime scaling issue you are describing) than we do for system-reserved. Hence this sentence in my post:

> We found that in order to safely guarantee node stability during high-load scenarios, we had to reserve a LOT more resources than baseline usage in our busier environments.

So in other words, we don't solve that scaling problem, because it's an emergent pattern of many underlying and unaddressed issues with reserved resources and the CPU manager (to say nothing of issues in containerd or the cri layer.) I was just describing how we cope with the problem.

**Comment by 2rs2ts** [comment]
Chiming in here again since I've been deeper in the guts of some of this stuff recently. The amount of `kubeReserved` currently used is actually way too low. 150mcpus for a node with up to 234 pods in one of our clusters, for example. Kubelet can EASILY gobble up way more CPU time than that. Forget containerd, I wonder how the current formula even came to be... our default on our Kubernetes the Hard Way clusters is to reserve 10% CPU as kubeReserved because that tends to capture at least the p75 of CPU spikes, and over there we're using the default `maxPods` of 110.

**Comment by stevehipwell** [comment]
@2rs2ts as covered in https://github.com/awslabs/amazon-eks-ami/issues/1141#issuecomment-1947048088 the kube reserved needs to be expressed as a function of pods and containers (Kubernetes scaling is multi-dimensional).

**Comment by 2rs2ts** [comment]
@stevehipwell absolutely, pod count is an important factor. The problem is that there is no magic formula because there are dozens of parameters, not just pod count. But putting aside the pursuit of perfection, we can still look at real-world data and conclude that the current reservations are _extremely_ lowballed, by an order of magnitude.

In real-world scenarios on our clusters, we hit way higher numbers than your tests did. I'll just give a worst-case from a node in our infra recently since (IMO) reservations should aim to account for the worst case (within reason.) Please note that it actually historically gets worse than this for us during peak hours, but these numbers are more than enough to show that the reservations are already far too low:

- `m7i.48xlarge` instance: 192cpus, 768GiB memory (744GiB after VM overhead)
- 74 pods
- 171 containers
- `/runtime.slice` usage over the past hour:
    - CPU max: 11.9 cores
    - CPU avg: 1.93 cores
    - Mem max: 1.10GiB
    - Mem avg: 966MiB
- `/system.slice` usage over the past hour:
    - CPU max: 4.11 cores
    - CPU avg: 1.23 cores
    - Mem max: 2.19GiB
    - Mem avg: 2.09GiB
- `/kubepods.slice` usage over the past hour (given for good measure, so you can form an idea of how not-that-busy this node is despite the very high runtime usage):
    - CPU max: 34.7 cores
    - CPU avg: 9.84 cores
    - Mem max: 209GiB
    - Mem avg: 37.6GiB

This is on an instance where we applied changes to the kubelet args to reserve much more than the default behavior, so there's breathing room. In our experience, doing so actually tends to lower the max usage numbers by a modest degree (probably because it reduces the resource contention caused by pods as a side effect.) I think the default behavior would have reserved 550mcpus on this type of instance, but I don't remember.

n.b.: there is a huge degree of variance in real-world patterns. Instances with 50% as many pods as my spot test have been observed using 300% as much resources for the runtime and the system. This is because of several factors, only some of which I can say for certain:

- High pod turnover/churn increases load on kubelet, perhaps more than the actual count at any given time
- Very large and busy pods without CPU limits will tend to gobble up resources from the rest of the system and cause things like PLEG remediation timeouts, which creates a sort of thrashing behavior in kubelet functionality
- Very high pod log output can cause more time spent on handling logs

**tl;dr** putting aside all else, the current calculations are at least an order of magnitude too low.

**Comment by stevehipwell** [comment]
You're preaching to the choir about the current spec being incorrect and there being many factors involved. But the important part of my comment was in regard to the containers not pods being of high significance.
